dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/export/changelog/ChangeLogServiceTest.java

```diff
@@ -0,0 +1,311 @@
+          assertEquals("25", changeLogs.get(0).change().dataValue().currentValue());
+          assertEquals("20", changeLogs.get(0).change().dataValue().previousValue());
+          assertEquals("20", changeLogs.get(1).change().dataValue().currentValue());
+          assertEquals("15", changeLogs.get(1).change().dataValue().previousValue());
+          assertEquals("15", changeLogs.get(2).change().dataValue().currentValue());
+          assertNull(changeLogs.get(2).change().dataValue().previousValue());
+        });
+  }
+
+  @Test
+  void shouldReturnChangeLogsWhenDataValueIsCreatedUpdatedAndDeleted()
+      throws IOException, NotFoundException {
+    injectSecurityContextUser(manager.get(User.class, "M5zQapPyTZI"));
+
+    Event event = manager.get(Event.class, "QRYjLTiJTrA");
+    assertNotNull(event);
+    String dataElementUid = event.getEventDataValues().iterator().next().getDataElement();
+    DataElement dataElement = manager.get(DataElement.class, dataElementUid);
+    assertNotNull(dataElement);
+
+    TrackerObjects trackerObjects = fromJson("tracker/event_and_enrollment.json");
+    updateDataValue(trackerObjects, event.getUid(), dataElementUid, "20");
+    assertNoErrors(trackerImportService.importTracker(importParams, trackerObjects));
+
+    updateDataValue(trackerObjects, event.getUid(), dataElementUid, "");
+    assertNoErrors(trackerImportService.importTracker(importParams, trackerObjects));
+
+    List<EventChangeLog> changeLogs = changeLogService.getEventChangeLog(UID.of("QRYjLTiJTrA"));
+
+    assertAll(
+        () -> {
+          validateNonValueElements(3, changeLogs, importUser, dataElementUid);
+          assertNull(changeLogs.get(0).change().dataValue().currentValue());
+          assertEquals("20", changeLogs.get(0).change().dataValue().previousValue());
+          assertEquals("20", changeLogs.get(1).change().dataValue().currentValue());
+          assertEquals("15", changeLogs.get(1).change().dataValue().previousValue());
+          assertEquals("15", changeLogs.get(2).change().dataValue().currentValue());
+          assertNull(changeLogs.get(2).change().dataValue().previousValue());
+        });
+  }
+
+  private void updateDataValue(
+      TrackerObjects trackerObjects, String eventUid, String dataElementUid, String newValue) {
+    trackerObjects.getEvents().stream()
+        .filter(e -> e.getEvent().equalsIgnoreCase(eventUid))
+        .findFirst()
+        .flatMap(
+            e ->
+                e.getDataValues().stream()
+                    .filter(
+                        dv -> dv.getDataElement().getIdentifier().equalsIgnoreCase(dataElementUid))
+                    .findFirst())
+        .ifPresent(dv -> dv.setValue(newValue));
+  }
+
+  private void validateNonValueElements(
+      int expectedListSize,
+      List<EventChangeLog> changeLogs,
+      User importUser,
+      String dataElementUid) {
+    assertEquals(expectedListSize, changeLogs.size());
```

Comment:
Since this is an integration tests we should be more careful with our assertions. Flaky tests in CI are annoying to fix with assertions that only state expected 3 got 2 😅 

Also you are using an `assertAll` where assertions that access an index lets say at 2 would cause an NPE in case the list only has 1 element.

You might need to rearrange your assertions.